### PR TITLE
🈵 Add an option to terminate the runtime when `memory.grow` fails

### PR DESCRIPTION
**Pull request by acfoltzer**
This new option determines whether to terminate the guest with a new `TerminationDetails::HeapOutOfMemory` variant when `memory.grow` fails, rather than returning `-1`. It is disabled by default, but can be set via `InstanceBuilder` or directly on the `Instance`.

This behavior deviates from the WebAssembly spec, but is useful in practice for determining when guest programs fail due to an exhausted heap.

Most languages will compile to code that includes an `unreachable` instruction if allocation fails, but this same instruction might also appear when other types of assertions fail, `panic!()` is called, etc. Terminating allows the error to be more directly identifiable.

### CODE REVIEW

File: lucet-runtime/lucet-runtime-internals/src/region.rs [hunk 1]
```diff
@@ -51,13 +51,7 @@ pub trait Region: RegionInternal {
...
 /// A `RegionInternal` is a collection of `Slot`s which are managed as a whole.
 pub trait RegionInternal: Send + Sync {
-    fn new_instance_with(
-        &self,
-        module: Arc<dyn Module>,
-        embed_ctx: CtxMap,
-        heap_memory_size_limit: usize,
-        alloc_strategy: AllocStrategy,
-    ) -> Result<InstanceHandle, Error>;
+    fn new_instance_with(&self, args: NewInstanceArgs) -> Result<InstanceHandle, Error>;
```
Comment (fst-crenshaw) [inline lucet-runtime/lucet-runtime-internals/src/region.rs:54]: Thanks for doing this added refactor.  It was time.

**Review by fst-crenshaw — APPROVED** [review]
Thanks so much for dropping everything today to do this urgent work.  I am grateful for your Good Decisions Under Pressure.  ❤️   

Concretely, I am grateful for your trio of tests, with especial appreciation for making sure the instance still works as expected when it stays within its heap bounds.  I am also grateful for the useful refactor you did to create `NewInstanceArgs` instead of adding yet another parameter to `new_instance_with`.

**Comment by acfoltzer** [comment]
Blocking this pending #584

**Review by fst-crenshaw — APPROVED** [review]
Thanks for addressing the UB that emerged from the recent addition of the [lucet_hostcall]!  Thanks for updating this PR!

**Pull request merged**